### PR TITLE
Add file io

### PR DESCRIPTION
**Pull request by perrymant**
1. This branch introduces the fileIO class, and can now read/write/append to/from file.
2. I had some git issues that I have hopefully sorted with this branch, including the readme.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+[![Build Status](https://travis-ci.org/perrymant/moneymaker.svg)](https://travis-ci.org/perrymant/moneymaker)
+![GitHub](https://img.shields.io/github/license/perrymant/moneymaker.svg)
+
+<pre>
```
Comment (kinbiko) [inline README.md]: `<pre>` is identical to triple backticks in markdown

File: src/main/java/io/github/perrymant/moneymaker/FileIO.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package io.github.perrymant.moneymaker;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+
+class FileIO {
+
+    static String read(String fileName) throws IOException {
```
Comment (kinbiko) [inline src/main/java/io/github/perrymant/moneymaker/FileIO.java]: No need for this to be `static`. In fact, other than `main`, try and avoid `static` unless you have a strong reason to use it.

Comment (kinbiko) [inline src/main/java/io/github/perrymant/moneymaker/FileIO.java]: If you've got `throws BlahException` in a method declaration that's a sign that you're doing something wrong. Remember: avoid checked exception - always wrap in (preferably your own implementation of a) RuntimeException.

File: src/main/java/io/github/perrymant/moneymaker/FileIO.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        BufferedReader text = new BufferedReader(new FileReader(fileName));
+        List<String> data = new ArrayList<>();
+        String s;
+        while ((s = text.readLine()) != null) {
```
Comment (kinbiko) [inline src/main/java/io/github/perrymant/moneymaker/FileIO.java]: There is [a lot easier way](https://stackoverflow.com/a/326440/2422278) of reading files than this. E.g.

```java
return new String(Files.readAllBytes(Paths.get(fileName)), Charset.defaultCharset());
```

File: src/main/java/io/github/perrymant/moneymaker/FileIO.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            data.add(s + "\n");
+        }
+        text.close();
+        return String.join("", data);
+    }
+
+    void write(String data, String fileName) throws IOException {
+        BufferedWriter writer = new BufferedWriter(new FileWriter(fileName));
+        writer.write(data);
+        writer.close();
```
Comment (kinbiko) [inline src/main/java/io/github/perrymant/moneymaker/FileIO.java]: Have a look at what try-with-resources or `finally` does in Java. Essentially we need to guarantee that `writer.close()` gets called, and if `writer.write(data)` throws an exception we end up not closing it.

File: src/main/java/io/github/perrymant/moneymaker/FileIO.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    void append(String dataToAppend, String fileName) throws IOException {
+        BufferedWriter writer = new BufferedWriter(new FileWriter(fileName, true));
```
Comment (kinbiko) [inline src/main/java/io/github/perrymant/moneymaker/FileIO.java]: magic boolean: What is `true`? Create a variable (or even a constant) so you can give it a name.

File: src/main/java/io/github/perrymant/moneymaker/Logger.java [hunk 6]
```diff
@@ -2,8 +2,19 @@
...
+                " | $$ $$/$$ $$ /$$__  $$| $$__  $$ /$$__  $$| $$  | $$| $$ $$/$$ $$ |____  $$| $$  /$$/ /$$__  $$ /$$__  $$ \n" +
+                " | $$  $$$| $$| $$  \\ $$| $$  \\ $$| $$$$$$$$| $$  | $$| $$  $$$| $$  /$$$$$$$| $$$$$$/ | $$$$$$$$| $$  \\__/ \n" +
+                " | $$\\  $ | $$| $$  | $$| $$  | $$| $$_____/| $$  | $$| $$\\  $ | $$ /$$__  $$| $$_  $$ | $$_____/| $$       \n" +
+                " | $$ \\/  | $$|  $$$$$$/| $$  | $$|  $$$$$$$|  $$$$$$$| $$ \\/  | $$|  $$$$$$$| $$ \\  $$|  $$$$$$$| $$       \n" +
+                " |__/     |__/ \\______/ |__/  |__/ \\_______/ \\____  $$|__/     |__/ \\_______/|__/  \\__/ \\_______/|__/       \n" +
+                "                                             /$$  | $$                                                      \n" +
+                "                                            |  $$$$$$/     ... Gold don't get old                           \n" +
+                "                                             \\______/                                                       ";
+        System.out.println(logo);
```
Comment (kinbiko) [inline src/main/java/io/github/perrymant/moneymaker/Logger.java]: Do we always want to print the logo whenever we want to log something in our application? I'm guessing no - maybe we should only print the logo once on startup?

File: src/test/java/io/github/perrymant/moneymaker/FileIOTest.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+
+public class FileIOTest {
+
+    @Test
+    public void canReadData() throws IOException {
+        String fileName = "src/test/java/io/github/perrymant/moneymaker/file_with_data";
```
Comment (kinbiko) [inline src/test/java/io/github/perrymant/moneymaker/FileIOTest.java]: sounds like this file should be in `src/test/resources` instead. Applies throughout

File: src/test/java/io/github/perrymant/moneymaker/FileIOTest.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        String target = FileIO.read(fileName);
+        assertEquals(expected, target);
+    }
+
+    @Test
+    public void canWriteData() throws IOException {
+        String fileName = "src/test/java/io/github/perrymant/moneymaker/file_with_data_write_test";
+        String dataToWrite = "2019/03/03, CREDIT, 200, 200, Got paid\n2019/03/04, DEBIT, 50, 150, Paid bill\n";
+        new FileIO().write(dataToWrite, fileName);
+        assertEquals(FileIO.read(fileName), dataToWrite);
```
Comment (kinbiko) [inline src/test/java/io/github/perrymant/moneymaker/FileIOTest.java]: Doesn't this kinda remove the need for the `canReadData` test?

File: src/test/java/io/github/perrymant/moneymaker/FileIOTest.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    @Test
+    public void canAppendData() throws IOException {
+        String fileName = "src/test/java/io/github/perrymant/moneymaker/file_with_data_append_test";
+        String dataToWrite = "2019/03/03, CREDIT, 200, 200, Got paid\n2019/03/04, DEBIT, 50, 150, Paid bill\n";
+        String dataToAppend = "2018/01/01, CREDIT, 200, 200, Got paid\n2018/01/02, DEBIT, 50, 150, Paid bill\n";
+        new FileIO().write(dataToWrite, fileName);
+        new FileIO().append(dataToAppend, fileName);
```
Comment (kinbiko) [inline src/test/java/io/github/perrymant/moneymaker/FileIOTest.java]: Doesn't this test remove the need for the `canWriteData()` test?

File: src/test/java/io/github/perrymant/moneymaker/file_with_data [hunk 10]
```diff
@@ -0,0 +1,2 @@
+2018/01/01, CREDIT, 200, 200, Got paid
```
Comment (kinbiko) [inline src/test/java/io/github/perrymant/moneymaker/file_with_data]: Should these files even be committed to Git if we only want to make sure we can read and write files? I suggest putting them in a `src/test/resources/dump/` folder, and then adding this folder to your `.gitignore`

**Review by kinbiko — CHANGES_REQUESTED** [review]
I see that you're adding file IO, and it looks like it works. But I don't see any requirements that say that we should do this? Can we update the README with this?

Let's go over what happened with the `master` branch when we meet up next.

Thanks for joining the slack btw!

**Comment by kinbiko** [comment]
Why was this one closed?

**Comment by perrymant** [comment]
... sorry, didn't know the procedure!
I thought I would re-state the situation in a new PR.

**Pull request closed without merging**